File: test/unit/loop_test.rb
#!/usr/bin/env ruby
$use_tree=false
#$use_tree=true

require_relative '../parser_test_helper'

class LoopTest < Test::Unit::TestCase #< ParserBaseTest <  EnglishParser

  include ParserTestHelper


  def _test_forever # OK ;{TRUST ME;}
    init 'beep forever'
    loops
    parse 'beep forever' # OK ;{TRUST ME;}
  end


  def test_loops  #OK
    parse 'beep three times' #OK
    parse "repeat three times: beep; okay" #OK
    parse "repeat three times: beep"       #OK
  end

  def test_try_until
    parse 'repeat while x<4: x++'
    assert_equals @variables[:x],4
    parse 'repeat x++ while x<4'
    assert_equals @variables[:x],4
    parse 'repeat x++ until x>4'
    assert_equals @variables[:x],5
    parse 'repeat until x>4: x++'
    assert_equals @variables[:x],5
    parse 'try until x>4: x++'
    assert_equals @variables[:x],5
    parse 'try while x<4: x++'
    assert_equals @variables[:x],4
    parse "try x++ until x>4"
    assert_equals @variables[:x],5
    parse "try x++ while x<4"
    assert_equals @variables[:x],4
    parse 'increase x until x>4'
    assert_equals @variables[:x],5
  end

  def test_every_date
    parse 'beep every three seconds'
    parse 'every three seconds make a beep'
  end

  def test_expressions
    #s "counter=0"
    #setter
    parse 'counter=1'
    #counter=@variables['counter']
    #@variables['counter']=1
    #parse "counter+1"
    #r=expression0
    assert(@variables['counter']==1)
    parse 'counter++'
    #r=expression0
    assert(@variables['counter']==2)
    #@variables['counter']=2
    parse 'counter+=1'
    #r=plusEqual
    #r=expression0
    assert(@variables['counter']==3)
    parse 'counter=counter+counter'
    #r=setter
    #r=algebra
    #r=expression0
    counter=@variables['counter']
    assert counter==6
  end

  def test_repeat # NEEEEDS blocks!! Parser.new(block)
    parse "counter =0; repeat three times: increase the counter; okay"
    assert_equals @variables['counter'],3
    assert_equals @variables[:counter],3
    assert "counter =3"
    #s "counter=counter+1;"
    #@interpret=false
    #action
    #@interpret=true
    parse 'counter =0; repeat three times: counter=counter+1; okay'
    assert 'counter =3' #if $use_tree # counter=counter+1 not repeatable as string
    parse 'counter =0; repeat three times: counter+=1; okay'
    assert 'counter =3'
    parse 'counter =0; repeat three times: counter++; okay'
    counter=@variables['counter']
    assert 'counter =3'
    #parse "counter =0; repeat three times: increase the counter by two; okay"
    #assert "counter =10"
  end

  def _test_forever # OK ;{TRUST ME;}
    @parser.s 'beep forever'
    @parser.loops
  end
end

File: angelos/test/loop_test.py
import _global
_global.use_tree = False
from parser_test_helper import *


class LoopTest(ParserBaseTest):
    

    def _test_forever(self):
        init('beep forever')
        loops()
        parse('beep forever')

    def test_while_return(self):
        assert_equals(parse('c=0;while c<1:c++;beep;done'), 'beeped')
        assert_equals(parse('c=0;while c<1:c++;beep;done'), 'beeped')

    def test_while_loop(self):
        parse('c=0;while c<3:c++;beep;done')
        assert(equals(3, variables['c']))

    def test_expressions(self):
        parse('counter=1')
        assert(equals(1, self.variableValues['counter']))
        parse('counter++')
        assert(equals(2, self.variableValues['counter']))
        parse('counter+=1')
        assert(equals(3, self.variableValues['counter']))
        parse('counter=counter+counter')
        counter = self.variableValues['counter']
        assert(counter.equals(6))

    def test_repeat(self):
        parse('counter =0; repeat three times: increase the counter; okay')
        assert('counter=3')
        assert_equals(self.variableValues['counter'], 3)

    def test_repeat3(self):
        assert_result_is('counter =0; repeat three times: counter=counter+1; okay', 3)
        assert_result_is('counter =0; repeat while counter < 3: counter=counter+1; okay', 3)

    def test_repeat1(self):
        parse('counter =0; repeat three times: counter+=1; okay')
        assert('counter =3')
        parse('counter =0; repeat three times: counter++; okay')
        counter = self.variableValues['counter']
        assert('counter =3')
        assert(counter.equals(3))

    def _test_forever(self):
        self.parser.s('beep forever')
        self.parser.loops()